numba/tests/test_dyn_array.py

```diff
@@ -0,0 +1,350 @@
+        arr.fill(123)  # test writability
+        np.testing.assert_equal(123, arr)
+        del arr
+
+    def test_empty_2d(self):
+        def pyfunc(m, n):
+            arr = np.empty((m, n), np.int32)
+            for i in range(m):
+                for j in range(n):
+                    arr[i, j] = i + j
+
+            return arr
+
+        cfunc = nrtjit(pyfunc)
+        m = 4
+        n = 3
+        expected_arr = pyfunc(m, n)
+        got_arr = cfunc(m, n)
+        np.testing.assert_equal(expected_arr, got_arr)
+
+        self.assertEqual(expected_arr.size, got_arr.size)
+        self.assertEqual(expected_arr.shape, got_arr.shape)
+        self.assertEqual(expected_arr.strides, got_arr.strides)
+
+        del got_arr
+
+    def test_empty_3d(self):
+        def pyfunc(m, n, p):
+            arr = np.empty((m, n, p), np.int32)
+            for i in range(m):
+                for j in range(n):
+                    for k in range(p):
+                        arr[i, j, k] = i + j + k
+
+            return arr
+
+        cfunc = nrtjit(pyfunc)
+        m = 4
+        n = 3
+        p = 2
+        expected_arr = pyfunc(m, n, p)
+        got_arr = cfunc(m, n, p)
+        np.testing.assert_equal(expected_arr, got_arr)
+
+        self.assertEqual(expected_arr.size, got_arr.size)
+        self.assertEqual(expected_arr.shape, got_arr.shape)
+        self.assertEqual(expected_arr.strides, got_arr.strides)
+
+        del got_arr
+
+    def test_empty_2d_sliced(self):
+        def pyfunc(m, n, p):
+            arr = np.empty((m, n), np.int32)
+            for i in range(m):
+                for j in range(n):
+                    arr[i, j] = i + j
+
+            return arr[p]
+
+        cfunc = nrtjit(pyfunc)
+        m = 4
+        n = 3
+        p = 2
+        expected_arr = pyfunc(m, n, p)
+        got_arr = cfunc(m, n, p)
+        np.testing.assert_equal(expected_arr, got_arr)
+
+        self.assertEqual(expected_arr.size, got_arr.size)
+        self.assertEqual(expected_arr.shape, got_arr.shape)
+        self.assertEqual(expected_arr.strides, got_arr.strides)
+
+        del got_arr
+
+    def test_return_global_array(self):
+        y = np.ones(4, dtype=np.float32)
+
+        def return_external_array():
+            return y
+
+        cfunc = nrtjit(return_external_array)
+        out = cfunc()
+
+        np.testing.assert_equal(y, out)
+        np.testing.assert_equal(y, np.ones(4, dtype=np.float32))
+        np.testing.assert_equal(out, np.ones(4, dtype=np.float32))
+
+    def test_return_global_array_sliced(self):
+        y = np.ones(4, dtype=np.float32)
+
+        def return_external_array():
+            return y[2:]
+
+        cfunc = nrtjit(return_external_array)
+        out = cfunc()
+
+        yy = y[2:]
+        np.testing.assert_equal(yy, out)
+        np.testing.assert_equal(yy, np.ones(2, dtype=np.float32))
+        np.testing.assert_equal(out, np.ones(2, dtype=np.float32))
+
+    def test_array_pass_through(self):
```

Comment:
Would it be meaningful to test that the refcount of the array is unchanged by calling the jitted function here?